// catapult-sdk/_build_browser/index.js
'use strict';

/*
 * Copyright (c) 2016-present,
 * Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
 *
 * This file is part of Catapult.
 *
 * Catapult is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Catapult is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Catapult.  If not, see <http://www.gnu.org/licenses/>.
 */

var auth = require('./auth/auth');

var _require = require('./crypto/keyPair'),
    createKeyPairFromPrivateKeyString = _require.createKeyPairFromPrivateKeyString,
    sign = _require.sign,
    verify = _require.verify;

var merkle = require('./crypto/merkleAuditProof');
var sha3Hasher = require('./crypto/sha3Hasher');
var address = require('./model/address');
var EntityType = require('./model/EntityType');
var idReducer = require('./model/idReducer');
var ModelType = require('./model/ModelType');
var networkInfo = require('./model/networkInfo');
var status = require('./model/status');
var serialize = require('./modelBinary/serialize');
var sizes = require('./modelBinary/sizes');
var transactionExtensions = require('./modelBinary/transactionExtensions');
var packetHeader = require('./packet/header');
var PacketType = require('./packet/PacketType');
var BinaryParser = require('./parser/BinaryParser');
var PacketParser = require('./parser/PacketParser');
var catapultModelSystem = require('./plugins/catapultModelSystem');
var BinarySerializer = require('./serializer/BinarySerializer');
var SerializedSizeCalculator = require('./serializer/SerializedSizeCalculator');
var arrayUtils = require('./utils/arrayUtils');
var base32 = require('./utils/base32');
var convert = require('./utils/convert');
var formattingUtils = require('./utils/formattingUtils');
var future = require('./utils/future');
var objects = require('./utils/objects');
var schemaFormatter = require('./utils/schemaFormatter');
var SchemaType = require('./utils/SchemaType');
var uint64 = require('./utils/uint64');

var catapultSdk = {
	auth: auth,
	constants: {
		sizes: sizes
	},
	crypto: {
		createKeyPairFromPrivateKeyString: createKeyPairFromPrivateKeyString,
		merkle: merkle,
		sha3Hasher: sha3Hasher,
		sign: sign,
		verify: verify
	},
	model: {
		address: address,
		EntityType: EntityType,
		idReducer: idReducer,
		ModelType: ModelType,
		networkInfo: networkInfo,
		status: status
	},
	modelBinary: {
		serialize: serialize,
		transactionExtensions: transactionExtensions
	},
	packet: {
		header: packetHeader,
		PacketType: PacketType
	},
	parser: {
		BinaryParser: BinaryParser,
		PacketParser: PacketParser
	},
	plugins: {
		catapultModelSystem: catapultModelSystem
	},
	serializer: {
		BinarySerializer: BinarySerializer,
		SerializedSizeCalculator: SerializedSizeCalculator
	},
	utils: {
		array: arrayUtils,
		base32: base32,
		convert: convert,
		formattingUtils: formattingUtils,
		future: future,
		objects: objects,
		schemaFormatter: schemaFormatter,
		SchemaType: SchemaType,
		uint64: uint64
	}
};

module.exports = catapultSdk;
