// catapult-sdk/_build_browser/utils/schemaFormatter.js
'use strict';

/*
 * Copyright (c) 2016-present,
 * Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
 *
 * This file is part of Catapult.
 *
 * Catapult is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Catapult is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Catapult.  If not, see <http://www.gnu.org/licenses/>.
 */

/** @module utils/schemaFormatter */
var SchemaType = require('./SchemaType');

// if 'definition' is a number, it is the type
// otherwise, it is an object with an optional type property (default type is none)
var getSchemaType = function getSchemaType(definition) {
	return 'number' === typeof definition ? definition : definition.type || SchemaType.none;
};

var getDefinition = function getDefinition(schema, key) {
	var definition = schema[key] || {};
	var type = getSchemaType(definition);

	return {
		type: type,
		resultKey: definition.resultKey || key,
		schemaName: definition.schemaName
	};
};

var getSchemaName = function getSchemaName(schemaName, entity) {
	return 'function' === typeof schemaName ? schemaName(entity) : schemaName;
};

/**
 * Formatter for formatting a catapult entity.
 * @class EntityFormatter
 */

/**
 * @function format
 * @param {object} entity The entity to format.
 * @returns {object} The formatted entity.
 * @memberof module:utils/schemaFormatter~EntityFormatter
 * @instance
 */

/** @exports utils/schemaFormatter */
var schemaFormatter = {
	/**
  * Formats an entity according to a schema and rules.
  * @param {object} entity The entity to format.
  * @param {object} entitySchema The schema corresponding to the entity.
  * @param {object} schemaDictionary A map of schema names to schemas for looking up component schemas.
  * @param {object} formattingRules A map for looking up formatting rules given a schema property type.
  * @returns {object} The formatted entity.
  */
	format: function format(entity, entitySchema, schemaDictionary, formattingRules) {
		// set object and array rules
		formattingRules[SchemaType.object] = function (value, format) {
			return format(value);
		};
		formattingRules[SchemaType.array] = function (array, format) {
			return array.map(function (value) {
				return format(value);
			});
		};
		formattingRules[SchemaType.dictionary] = function (dictionary, format) {
			var result = {};
			Object.keys(dictionary).forEach(function (key) {
				result[key] = format(dictionary[key]);
			});

			return result;
		};

		var result = {};
		Object.keys(entity).forEach(function (key) {
			var definition = getDefinition(entitySchema, key);
			var rule = formattingRules[definition.type];
			if (rule) {
				var schemaName = getSchemaName(definition.schemaName, entity[key]);
				result[definition.resultKey] = rule(entity[key], 'number' === typeof schemaName // if schemaName is a number, interpret it as a rule id and format child parts as values
				? formattingRules[schemaName] : function (value) {
					return schemaFormatter.format(value, schemaDictionary[schemaName], schemaDictionary, formattingRules);
				});
			}
		});

		return result;
	}
};

module.exports = schemaFormatter;
