// catapult-sdk/_build_browser/plugins/lock.js
'use strict';

/*
 * Copyright (c) 2016-present,
 * Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
 *
 * This file is part of Catapult.
 *
 * Catapult is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Catapult is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Catapult.  If not, see <http://www.gnu.org/licenses/>.
 */

/** @module plugins/lock */
var EntityType = require('../model/EntityType');
var ModelType = require('../model/ModelType');
var sizes = require('../modelBinary/sizes');

var constants = { sizes: sizes };

var commonLockInfoSchema = {
	account: ModelType.binary,
	accountAddress: ModelType.binary,
	mosaicId: ModelType.uint64,
	amount: ModelType.uint64,
	height: ModelType.uint64
	/* status */
};

var commonLockTransactionSchema = {
	mosaicId: ModelType.uint64,
	amount: ModelType.uint64,
	duration: ModelType.uint64
};

var parseLockData = function parseLockData(parser, transaction) {
	transaction.mosaicId = parser.uint64();
	transaction.amount = parser.uint64();
	transaction.duration = parser.uint64();
};

var serializeLockData = function serializeLockData(transaction, serializer) {
	serializer.writeUint64(transaction.mosaicId);
	serializer.writeUint64(transaction.amount);
	serializer.writeUint64(transaction.duration);
};

/**
 * Creates a lock plugin.
 * @type {module:plugins/CatapultPlugin}
 */
var lockPlugin = {
	registerSchema: function registerSchema(builder) {
		builder.addSchema('hashLockInfo', {
			lock: { type: ModelType.object, schemaName: 'hashLockInfo.lock' }
		});
		builder.addSchema('hashLockInfo.lock', Object.assign({}, commonLockInfoSchema, {
			hash: ModelType.binary
		}));

		builder.addSchema('secretLockInfo', {
			lock: { type: ModelType.object, schemaName: 'secretLockInfo.lock' }
		});
		builder.addSchema('secretLockInfo.lock', Object.assign({}, commonLockInfoSchema, {
			/* hashAlgorithm */
			secret: ModelType.binary,
			recipient: ModelType.binary
		}));

		builder.addTransactionSupport(EntityType.hashLock, Object.assign({}, commonLockTransactionSchema, {
			hash: ModelType.binary
		}));
		builder.addTransactionSupport(EntityType.secretLock, Object.assign({}, commonLockTransactionSchema, {
			secret: ModelType.binary,
			recipient: ModelType.binary
		}));
		builder.addTransactionSupport(EntityType.secretProof, {
			secret: ModelType.binary,
			proof: ModelType.binary
		});
	},

	registerCodecs: function registerCodecs(codecBuilder) {
		codecBuilder.addTransactionSupport(EntityType.hashLock, {
			deserialize: function deserialize(parser) {
				var transaction = {};
				parseLockData(parser, transaction);

				transaction.hash = parser.buffer(constants.sizes.hash256);
				return transaction;
			},

			serialize: function serialize(transaction, serializer) {
				serializeLockData(transaction, serializer);

				serializer.writeBuffer(transaction.hash);
			}
		});

		codecBuilder.addTransactionSupport(EntityType.secretLock, {
			deserialize: function deserialize(parser) {
				var transaction = {};
				parseLockData(parser, transaction);

				transaction.hashAlgorithm = parser.uint8();
				transaction.secret = parser.buffer(constants.sizes.hash512);
				transaction.recipient = parser.buffer(constants.sizes.addressDecoded);
				return transaction;
			},

			serialize: function serialize(transaction, serializer) {
				serializeLockData(transaction, serializer);

				serializer.writeUint8(transaction.hashAlgorithm);
				serializer.writeBuffer(transaction.secret);
				serializer.writeBuffer(transaction.recipient);
			}
		});

		codecBuilder.addTransactionSupport(EntityType.secretProof, {
			deserialize: function deserialize(parser) {
				var transaction = {};
				transaction.hashAlgorithm = parser.uint8();
				transaction.secret = parser.buffer(constants.sizes.hash512);
				var proofSize = parser.uint16();
				transaction.proof = parser.buffer(proofSize);
				return transaction;
			},

			serialize: function serialize(transaction, serializer) {
				serializer.writeUint8(transaction.hashAlgorithm);
				serializer.writeBuffer(transaction.secret);
				var proofSize = transaction.proof.length;
				serializer.writeUint16(proofSize);
				serializer.writeBuffer(transaction.proof);
			}
		});
	}
};

module.exports = lockPlugin;
