// catapult-sdk/_build_browser/plugins/multisig.js
'use strict';

/*
 * Copyright (c) 2016-present,
 * Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
 *
 * This file is part of Catapult.
 *
 * Catapult is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Catapult is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Catapult.  If not, see <http://www.gnu.org/licenses/>.
 */

/** @module plugins/multisig */
var EntityType = require('../model/EntityType');
var ModelType = require('../model/ModelType');
var sizes = require('../modelBinary/sizes');
var convert = require('../utils/convert');

var constants = { sizes: sizes };

/**
 * Creates a multisig plugin.
 * @type {module:plugins/CatapultPlugin}
 */
var multisigPlugin = {
	registerSchema: function registerSchema(builder) {
		builder.addTransactionSupport(EntityType.modifyMultisigAccount, {
			modifications: { type: ModelType.array, schemaName: 'modifyMultisigAccount.modification' }
		});
		builder.addSchema('modifyMultisigAccount.modification', {
			cosignatoryPublicKey: ModelType.binary
		});

		builder.addSchema('multisigEntry', {
			multisig: { type: ModelType.object, schemaName: 'multisigEntry.multisig' }
		});
		builder.addSchema('multisigEntry.multisig', {
			account: ModelType.binary,
			accountAddress: ModelType.binary,
			multisigAccounts: { type: ModelType.array, schemaName: ModelType.binary },
			cosignatories: { type: ModelType.array, schemaName: ModelType.binary }
		});
		builder.addSchema('multisigGraph', {
			multisigEntries: { type: ModelType.array, schemaName: 'multisigEntry' }
		});
	},

	registerCodecs: function registerCodecs(codecBuilder) {
		codecBuilder.addTransactionSupport(EntityType.modifyMultisigAccount, {
			deserialize: function deserialize(parser) {
				var transaction = {};
				transaction.minRemovalDelta = convert.uint8ToInt8(parser.uint8());
				transaction.minApprovalDelta = convert.uint8ToInt8(parser.uint8());

				var numModifications = parser.uint8();
				transaction.modifications = [];

				while (transaction.modifications.length < numModifications) {
					var type = parser.uint8();
					var cosignatoryPublicKey = parser.buffer(constants.sizes.signer);
					transaction.modifications.push({ type: type, cosignatoryPublicKey: cosignatoryPublicKey });
				}

				return transaction;
			},

			serialize: function serialize(transaction, serializer) {
				serializer.writeUint8(convert.int8ToUint8(transaction.minRemovalDelta));
				serializer.writeUint8(convert.int8ToUint8(transaction.minApprovalDelta));

				var numModifications = transaction.modifications.length;
				serializer.writeUint8(numModifications);

				transaction.modifications.forEach(function (modification) {
					serializer.writeUint8(modification.type);
					serializer.writeBuffer(modification.cosignatoryPublicKey);
				});
			}
		});
	}
};

module.exports = multisigPlugin;
